usr/tgtd.c

```diff
@@ -519,6 +535,203 @@ static int parse_params(char *name, char *p)
+	json_auto_t *root = json_loads(data, 0, &error);
+
+	free(data);
+	if (root == NULL) {
+		set_err_msg(resp, TGT_ERR_INVALID_JSON,
+			"json config is incorrect");
+		return HA_CALLBACK_CONTINUE;
+	}
+
+	json_t *tname = json_object_get(root, "TargetName");
+	if (!json_is_string(tname)) {
+		set_err_msg(resp, TGT_ERR_INVALID_TARGET_NAME,
+			"TargetName is not string");
+		return HA_CALLBACK_CONTINUE;
+	}
+
+	memset(cmd, 0, sizeof(cmd));
+	snprintf(cmd, sizeof(cmd), "tgtadm --lld iscsi --mode target --op new"
+		" --tid=%s --targetname=%s", tid, json_string_value(tname));
+	rc = exec(cmd);
+
+	if (rc) {
+		set_err_msg(resp, TGT_ERR_TARGET_CREATE,
+			"target create failed");
+		return HA_CALLBACK_CONTINUE;
+	}
+	ha_set_empty_response_body(resp, HTTP_STATUS_OK);
+
+	return HA_CALLBACK_CONTINUE;
+}
+
+static int lun_create(const _ha_request *reqp,
+	_ha_response *resp, void *userp)
+{
+	char cmd[512];
+	const char *tid = ha_parameter_get(reqp, "tid");
+	const char *lid = ha_parameter_get(reqp, "lid");
+	int rc = 0;
+	char *data = NULL;
+
+	if (tid == NULL) {
+		set_err_msg(resp, TGT_ERR_INVALID_PARAM,
+			"tid param not given");
+		return HA_CALLBACK_CONTINUE;
+	}
+
+	if (lid == NULL) {
+		set_err_msg(resp, TGT_ERR_INVALID_PARAM,
+			"lid param not given");
+		return HA_CALLBACK_CONTINUE;
+	}
+
+	data = ha_get_data(reqp);
+	if (data == NULL) {
+		set_err_msg(resp, TGT_ERR_NO_DATA,
+			"json config not given");
+		return HA_CALLBACK_CONTINUE;
+	}
+
+	json_error_t error;
+	json_auto_t *root = json_loads(data, 0, &error);
+
+	free(data);
+	if (root == NULL) {
+		set_err_msg(resp, TGT_ERR_INVALID_JSON,
+			"json config is incorrect");
+		return HA_CALLBACK_CONTINUE;
+	}
+
+	json_t *dev_path = json_object_get(root, "DevPath");
+	if (!json_is_string(dev_path)) {
+		set_err_msg(resp, TGT_ERR_INVALID_LUN_PATH,
+			"DevPath is not string");
+		return HA_CALLBACK_CONTINUE;
+	}
+
+	json_t *vmid = json_object_get(root, "VmID");
+	if (!json_is_string(vmid)) {
+		set_err_msg(resp, TGT_ERR_INVALID_VMID,
+			"VmID is not string");
+		return HA_CALLBACK_CONTINUE;
+	}
+
+	json_t *vmdkid = json_object_get(root, "VmdkID");
+	if (!json_is_string(vmdkid)) {
+		set_err_msg(resp, TGT_ERR_INVALID_VMDKID,
+			"VmdkID is not string");
+		return HA_CALLBACK_CONTINUE;
+	}
+
+	memset(cmd, 0, sizeof(cmd));
+	snprintf(cmd, sizeof(cmd), "tgtadm --lld iscsi --mode logicalunit --op new"
+		" --tid=%s --lun=%s -b %s --bstype hyc --bsopts vmid=%s:vmdkid=%s",
+		tid, lid, json_string_value(dev_path), json_string_value(vmid),
+		json_string_value(vmdkid));
+	rc = exec(cmd);
+
+	if (rc) {
+		set_err_msg(resp, TGT_ERR_LUN_CREATE,
+			"target create failed");
+		return HA_CALLBACK_CONTINUE;
+	}
+	ha_set_empty_response_body(resp, HTTP_STATUS_OK);
+
+	return HA_CALLBACK_CONTINUE;
+}
+
+int tgt_ha_start_cb(const _ha_request *reqp,
+	_ha_response *resp, void *userp)
+{
+	struct _ha_instance *hap = (struct _ha_instance *) userp;
+
+	pthread_create(&ha_hb_tid, NULL, &ha_heartbeat, (void *)hap);
```

Comment:
Check return value. What if thread didn't start?

Reply:
What happens when this can is called multiple times?

Reply:
return value checked, if failed, we return error. handled multiple times occurence too.